### ISSUE TITLE
Error: the repeated events will not be showed in the list of future events only once when the option Shortened view is checked

### ISSUE THREAD

**Issue by jenda77** [contao/core#8497]
Unfortunately, the repeated events will not be showed in the list of future events only once when the option Shortened view is checked.

**Comment by leofeyer** [comment]
See #8447.

**Comment by Aybee** [comment]
This is correct behaviour. `Shortened view` regards to events that span multiple days only, not to recurring events.
> Tooltip: Show events only once even if they span multiple days.

**Comment by fritzmg** [comment]
@Aybee: I think you misunderstood. When an event spans multiple days, there will be an entry for that event for each day in the event list. If you activate the option _Shortened view_, you will see only one entry on the first day of the event in the event list.

The bug here is, that if the event is also recurring, it will _never_ show up in the event list again after the first day of the given date in the event list.

**Comment by jenda77** [comment]
@leofeyer No, this is not what I mean. I mean, when I choose the option Shortened view I expect that the event will be showed only once and not for every single day when it takes part again.

**Comment by fritzmg** [comment]
@jenda77 I cannot reproduce this in the Contao Online Demo. Do you have reproduction steps?

**Comment by jenda77** [comment]
This is my template for the event list

`<div class="event layout_list<?php echo $this->classList; ?>">
<?php if ($this->header): ?>
  <div class="header<?php echo $this->classHeader; ?>">
<?php if ($this->daysElse): ?>
<span class="date"><?php echo $this->daysElse ?></span>
<?php else: ?>
<span class="date"><?php echo ($this->parseDate("d.m.", $this->startDate)); ?><?php if ($this->endDate): ?> - <?php echo ($this->parseDate("d.m.", $this->endDate)); ?><?php endif ?></span><?php endif ?><?php if ($this->timesDaily): ?>
<span class="time"><?php echo $this->timesDaily ?></span>
<?php else: ?>
<?php if ($this->startTime): ?>
<span class="time"> <?php echo ($this->parseDate("H:i", $this->startTime)); ?><?php if ($this->endTime): ?> - <?php echo ($this->parseDate("H:i", $this->endTime)); ?><?php endif ?></span><?php endif ?>
<?php endif ?>
  <?php if ($this->location): ?>
 <span class="location"><i class="fa fa-map-marker" aria-hidden="true"></i> <?php echo $this->location ?> </span> <?php endif ?>
 
  </div>
<?php endif; ?>
  <h2 <?php if ($this->subtitle): ?>class="with-subtitle"<?php endif ?>><a href="<?php echo $this->href; ?>" title="<?php echo $this->title; ?> (<?php if ($this->day): echo $this->day; ?>, <?php endif; echo $this->date; if ($this->time): ?>, <?php echo $this->time; endif; ?>)"<?php echo $this->target; ?>><?php echo $this->link; ?></a></h2>
  <?php if ($this->subtitle): ?>
<h3><?php echo $this->subtitle ?></h3>
<?php endif ?>
  <?php if ($this->addImage): ?>
    <figure class="image_container<?php echo $this->floatClass; ?>"<?php if ($this->margin || $this->float): ?> style="<?php echo trim($this->margin . $this->float); ?>"<?php endif; ?>>

      <?php if ($this->href): ?>
        <a href="<?php echo $this->href; ?>"<?php echo $this->attributes; ?> title="<?php echo $this->alt; ?>">
      <?php endif; ?>

      <?php $this->insert('picture_default', $this->picture); ?>

      <?php if ($this->href): ?>
        </a>
      <?php endif; ?>

      <?php if ($this->caption): ?>
        <figcaption class="caption" style="width:<?php echo $this->arrSize[0]; ?>px"><?php echo $this->caption; ?></figcaption>
      <?php endif; ?>

    </figure>
  <?php endif; ?>
  <?php if ($this->teaser): ?>
<div class="teaser"><?php echo $this->teaser; ?>
<?php if ($this->eventConnection): ?>
<p><span class="connection"><i class="fa fa-subway" aria-hidden="true"></i> <?php echo $this->eventConnection ?></span></p>
<?php endif ?>
</div>
<?php endif ?>
  <?php if ($this->details): ?>
    <p class="more"><i class="fa fa-angle-double-right" aria-hidden="true"></i> <a href="<?php echo $this->href; ?>" title="<?php echo $this->readMore; ?>"<?php echo $this->target; ?>><?php echo $this->more; ?> <span class="invisible"><?php echo $this->link; ?></span></a></p>
  <?php endif; ?>  
	<?php if ($this->showTags): ?>
	<?php if (count($this->taglist)): ?>
	<ul class="tag-chain">
	<?php $counter = 0; foreach ($this->taglist as $tag): ?>
	<li class="tag-chain-item<?php if ($counter == 0) echo ' first'; ?><?php if ($counter == count($this->taglist)-1) echo ' last'; ?><?php if ($this->showTagClass) echo ' ' . $tag['class']; ?>"><?php echo $tag['url']; ?></li>
	<?php $counter++; endforeach; ?>
	</ul>
	<?php endif; ?>
	<?php endif; ?>
</div>`

And the screenshot of my settings for the module I use for the frontend list. 

<img width="913" alt="snimek obrazovky 2016-09-21 v 7 59 16" src="https://cloud.githubusercontent.com/assets/3715050/18708178/e0fc898c-7ff9-11e6-87de-67e336ed4983.png">

**Comment by fritzmg** [comment]
You should provide reproduction steps that can be done in the Contao Online Demo.

The template should not matter. If it happens only with a custom template, then it is not a bug.

**Comment by jenda77** [comment]
The steps are: 

1. I´ve created the modules for event list and event reader
2. after this I´ve placed the module for the event list on the site where I need it.
 
That´s it.

**Comment by fritzmg** [comment]
@jenda77 

* you did not include the settings of the event list module
* you did not include what date you set for your event
* you did not include how you are viewing the event list in the frontend

**Comment by fritzmg** [comment]
As already said, you should provide __detailed__ reproduction stept __for the Contao Online Demo__. _You need to do them yourself first there_, so that you can confirm for yourself, that it also occurs in the Contao Online Demo.

**Comment by jenda77** [comment]
Sorry, it´s obviously that I did it. ;-) The settings for the event list you can see above.

**Comment by fritzmg** [comment]
How did you reproduce it in the Contao Online Demo?

**Comment by fritzmg** [comment]
When I set the event "Contao North Day" to start at the `26.10.2014` and end at the `28.10.2014` and set it to recurring each year and the event list module to _Shortened view_, everything looks correct in the frontend:

![screen shot 2016-09-21 at 13 19 01](https://cloud.githubusercontent.com/assets/4970961/18709030/39006dde-7ffe-11e6-8f2e-76ea7bf1baf4.png)

**Comment by fritzmg** [comment]
However, I am noticing another problem in the screenshot: when there is a start and end __time__, only the time will be shown and no date at all. I am not sure if that's intentional?

**Comment by Aybee** [comment]
> @fritzmg The bug here is, that if the event is **also** recurring, it will never show up in the event list again after the first day of the given date in the event list.

I know the behaviour but it looks like I missed the word **also** in your description which I did not test.

But I think you did the test know without finding a bug.

**Comment by jenda77** [comment]
Sorry, I don't have any energy more for contao now. My explanation is above, more I don't want to describe. 


Jan Kout

--------------------------
DE: +49 176 457 99 488
CZ: +420 728 781 067

www.jankout.eu
jendakout@gmail.com

21. 9. 2016 v 13:33, Andreas Burg <notifications@github.com>:

@fritzmg The bug here is, that if the event is also recurring, it will never show up in the event list again after the first day of the given date in the event list.

I know the behaviour but it looks like I missed the word also in your description which I did not test.

But I think you did the test know without finding a bug.

**Comment by fritzmg** [comment]
> My explanation is above, more I don't want to describe.

I cannot reproduce your issue based on your explanation. See the screenshot above.

**Comment by jenda77** [comment]
Just ignore it. Thank you for help.

**Comment by Aybee** [comment]
@jenda77 
Melde dich einfach im Forum nochmal, vielleicht können wir das dort ja erstmal in Ruhe abklären.

**Comment by leofeyer** [comment]
What are the exact steps to reproduce this in the online demo?

**Comment by jenda77** [comment]
I change the date for the first event: start date 21.9.2016 end date 25.9.2016 and disable the option Repeat event. Then this event does not appear in the list of upcoming events. But it still takes place. 

<img width="937" alt="snimek obrazovky 2016-09-23 v 13 43 18" src="https://cloud.githubusercontent.com/assets/3715050/18784710/1b03ba06-8194-11e6-8f1d-aff8223805c0.png">
<img width="947" alt="snimek obrazovky 2016-09-23 v 13 43 28" src="https://cloud.githubusercontent.com/assets/3715050/18784712/212d1ef4-8194-11e6-9fdf-f523ebce2d42.png">
<img width="729" alt="snimek obrazovky 2016-09-23 v 13 43 35" src="https://cloud.githubusercontent.com/assets/3715050/18784715/275d8d5e-8194-11e6-84cb-75b49856b554.png">

**Comment by Aybee** [comment]
Same regards to one-day events with `Start time` and `End time`.

Maybe we need a checkbox in the List "Handle running events as upcoming events (not bygone)". Then match them against endDateTime but not startDateTime. So a running events will not become `bygone`.

**Comment by jenda77** [comment]
I think we don´t need more options. Just the right (linked to the reality) handling of events. The logic is that when an event has en end date in the future but the start day in the past should still be showed in the list of upcoming list.

**Comment by asaage** [comment]
@jenda77 That is your reality...
for most of my Events that would not make any sense.
So I'd favour the option as described by @Aybee.

**Comment by jenda77** [comment]
But when the event still takes place you want to have it in the list of upcoming events. Or not?

**Comment by asaage** [comment]
I'd want the events in the list for other people to join in.
But i definitely dont want people to show up in the middle of the event.
That depends of course on the nature of the event ;-)

**Comment by jenda77** [comment]
OK, in my experience it´s common in the culture scene how Hebbel am Ufer has it [www.hebbel-am-ufer.de](http://www.hebbel-am-ufer.de/programm/)). They show it either as repeated event for each day particularly or, if they want, as one event of several days.

**Comment by Physiklehrer** [comment]
My Forum thread about this issue:
https://community.contao.org/de/showthread.php?63875-3-5-16-Laufende-Events-werden-nicht-mehr-unter-quot-upcoming-quot-angezeigt

Online Demo --> create an event starting yesterday and ending tomorrow --> event is currently running, but not displayed in the "upcoming events example" (since 3.5.16, before it was). This might be ok, if it is decision that "upcoming" excludes running events. Unfortunately there is no possibilty to show currently running events any more. Also the event list format "day" does not show this running event, output: "There are no events on this day."

Otherwise, currently running events are shown at the event list format "past events". But they are still running and not "past" -  if we want to be that strict with "upcoming"?! Here my suggestion to make the situation more clearly:

Solution 1: both "upcoming" and "past events" INCLUDE all currently running events (like contao <= 3.5.15)
- Programming: past events are displayed if system date/time < END date/time, upcoming events are displayed if system date/time > START date/time.
- all users who don't want do show running events in the past or upcoming list (see "But i definitely don´t want people to show up in the middle of the event.") can realize that with "Show from" or "Show until" in the "Publish settings"-section.
- interpretation: running events are both upcoming and past, because in the middle of a running event it is 1/2 "past" and 1/2 "upcome".

Solution 2: "past" and "upcoming" events EXCLUDE currently running events
- programming: past events are only displayed if system date / time > END date / time, upcoming events are displayed if system date / time < START date / time
- Interpretation: there are no 1/2 or 3/4... events. An Event is past, running or upcoming without any intersection.

In this case, a way to show currently running events is needed. Ideas:

2a) add further items to "event list format": "past", "upcoming", "past and running", "upcoming and running"
2b) add a checkbox "show / include running events" etc. to the event list module (proposal of Aybee above)
2c) add a checkbox "show event while running" to the event properties, so that it can be speficied for each event (proposal of tab in the forum)

For me it doesn't matter which solution is realised, take the one that is easy to implement (without getting new bugs with repeating events, I understand that this is not easy... many thanks to the core team!). Anyway, in my opinion this is a defect, not a new feature, because contao <=3.5.15 was able to show running events.

**Comment by asaage** [comment]
2a) does not seem to be practicable as for "past" and "upcoming" there is already a long list for the different timespans
2c) having this as a property of the event itself might prevent you from displaying in alternative listings

so imho 2b) should be the way to go
additional checkbox for eventlist-Module like
- [ ] __take running events into account__

or 
- [ ] __discard running events__

![image](https://cloud.githubusercontent.com/assets/4138336/18963445/4c5a0930-8675-11e6-9047-739dc0a86a5f.png)

**Comment by fritzmg** [comment]
New features won't get added to Contao 3.5. The old behaviour should be reinstated - if that was the way it always worked in all other Contao 3 versions. If the latter is the case, the current behaviour constitutes a BC break.

**Comment by asaage** [comment]
> New features won't get added to Contao 3.5.

maybe new bugs shouldn't get added to 3.5. as well 😄 
nevertheless it would be a nice feature for 4.x

**Comment by jenda77** [comment]
@Physiklehrer thank you. This is what I mean all the time. It's not correct handling of the event module. Before the current version it worked well. 

@leofeyer please, reopen this issue, it's a bug. And, please, fix it. Thank you very much.

**Comment by fritzmg** [comment]
Indeed, it's not a duplicate of #8447

**Comment by jenda77** [comment]
@leofeyer Kannst du das bitte wieder öffnen? Es ist nämlich ein Fehlverhalten des Eventsmoduls. Danke.

**Comment by Physiklehrer** [comment]
The problem is that the "// Skip occurrences in the past"-section on line 176 in ModuleEventlist.php is also executed if there is no repeating event. Solution with an additional if-section:

`		// Remove events outside the scope
		foreach ($arrAllEvents as $key=>$days)
		{
			foreach ($days as $day=>$events)
			{
				foreach ($events as $event)
				{
					// Use repeatEnd if > 0 (see #8447)
					if ($event['repeatEnd'])
					{
						if ($event['repeatEnd'] < $intStart || $event['startTime'] > $intEnd)
						{
							continue;
						}
					
						// Skip occurrences in the past
						if (strtotime($event['datetime']) < $intStart)
						{
							continue;
						}
					} else
					{
						if ($event['endTime'] < $intStart || $event['startTime'] > $intEnd)
						{
							continue;
						}
					}

					$event['firstDay'] = $GLOBALS['TL_LANG']['DAYS'][date('w', $day)];
					$event['firstDate'] = \Date::parse($objPage->dateFormat, $day);

					$arrEvents[] = $event;
				}
			}
		}`

**Comment by Physiklehrer** [comment]
PS: with my solution above running single events are displayed, but no running repeated events. To do that you should check the ENDtime of the repeated event at "// Skip occurrences in the past", not "datetime", e.g. `if (strtotime($event['ENDtime']) < $intStart)`. But that doesn't work, the variable needs to be created first...

**Comment by fritzmg** [comment]
Looks good, may be you can create a pull request for that.

**Comment by frank-mueller** [comment]
Gleiches Problem sowohl bei 3.5.17 (demo.contao.org) als auch bei 4.2.4 (lokal). Vorgehensweise hab ich hier https://community.contao.org/de/showthread.php?63397-Event-in-verk%FCrzter-Darstellung-bis-zu-letzten-Tag-zeigen&p=422262#post422262 im Post 9 dargestellt.

**Comment by Physiklehrer** [comment]
ok, die if-Verzweigung braucht es natürlich eigentlich nicht, sorry. Statt meiner obigen unnötig komplizierten Lösung würde es auch genügen, einfach Zeile 175 durch `if ($event['repeatEnd'] && strtotime($event['datetime']) < $intStart)` zu ersetzen...

Ich versuche mich dann mal an einem Pull-Request, auch wenn das leider noch nicht ganz perfekt ist, da laufende Wiederholungsevents weiterhin ausgeblendet werden. Angesichts der zunehmend Zahl von Anfragen im Forum hierzu sollte zumindest dies aber unbedingt in die 2.5.18...

**Comment by leofeyer** [comment]
Das Grundproblem ist, wie @asaage es schon oben beschrieben hat, dass es von dem Event abhängt, ob er während der Laufzeit noch "upcoming" ist oder nicht.

Eine Kirmes, die 5 Tage lang dauert, kann ich natürlich noch am dritten Tag besuchen, daher sollte das Event angezeigt werden. Hingegen kann an einer Urlaubsreise, bei der der Flieger am ersten Tag geht, am dritten Tag nicht mehr teilgenommen werden, daher sollte das Event nicht mehr angezeigt werden.

Wir hatten genau diesen Fall im letzten Mumble-Call mit den @contao/developers diskutiert und die Lösung wäre, dass man pro Event auswählen kann, ob er "offen" oder "geschlossen" ist (die Wortwahl steht noch nicht fest).

Für Contao 3 wird es wohl keine Änderung des Verhaltens mehr geben (siehe #153, #8194 und #8447) aber für Contao 4 wäre so ein Feature denkbar.

**Comment by fritzmg** [comment]
> Für Contao 3 wird es wohl keine Änderung des Verhaltens mehr geben (siehe #153, #8194 und #8447) aber für Contao 4 wäre so ein Feature denkbar.

Sollte für Contao 3 dann nicht das Original Verhalten aus Contao 3.0.0 bis 3.5.12 wieder hergestellt werden?

**Comment by asaage** [comment]
> und die Lösung wäre, dass man pro Event auswählen kann, ob er "offen" oder "geschlossen"

pro Event fände ich das nicht so gut - ich finde es sollte im Listenmodul festgelegt werden.

**Comment by leofeyer** [comment]
> ich finde es sollte im Listenmodul festgelegt werden

Das geht nicht. Nimm die beiden Beispiel-Events (Kirmes und Urlaubsreise) und stell Dir vor, sie wären beide im selben Kalender. Wie willst Du im Listenmodul sagen, dass der eine Event "offen" ist und der andere "geschlossen"?

**Comment by Physiklehrer** [comment]
Für den Fall "Urlaubsreisen" besteht derzeit aber schon die Möglichkeit, durch Eintragen des Startdatums in das "Anzeigen bis"-Feld das Event nach dem Start auszublenden, um z.B. eine Anmeldung zu verhindern. Natürlich verschwindet es dann auch aus Kalendern und Listen mit vergangenen Events, doch bei Events sind die Rückblicke sicher nicht ganz so elementar wie zukünftige Events.

Das Kirmes-Beispiel lässt sich hingegen derzeit nicht mehr mit Core-Features realisieren, da laufende Events wegen des "// Skip occurrences in the past"-Abschnitts grundsätzlich ausgeblendet werden. Die Core-Features decken jetzt also weniger Anwendungsfälle ab als vor 3.5.12. Von daher würde ich schon vorschlagen, meinen kleinen Fix einzubauen, sodass zumindest laufende Einzelevents wieder wie früher angezeigt werden.

**Comment by Physiklehrer** [comment]
Für das ggf. neue Feature in Contao 4:
> pro Event fände ich das nicht so gut - ich finde es sollte im Listenmodul festgelegt werden.

Das sehe ich auch so, man kann ja einen Kalender "Kirmes" ("offen") und einen "Urlaubsreisen" ("geschlossen") anlegen und im Modul dann beide anzeigen lassen, wenn man eine gemischte Ausgabe will. Wenn man das für jedes Event extra machen muss ist das vermutlich fehleranfälliger.

Allerdings frage ich mich auch, ob es diese zusätzliche Option wirklich braucht, so lange der Core selbst keine Event-Anmeldungen unterstützt. Welcher Anbieter von Urlaubsreisen will, dass man sich noch 1 Minute vor Abflug anmelden kann? Das Häkchen "geschlossen" reicht dann sowieso nicht, es wird eine Datumsangabe wie bei "Anzeigen bis" benötigt. Und das könnte dann auch die entsprechende Erweiterung, die die Event-Anmeldung liefert, mitbringen (tut das calendar_extended nicht schon?!)

**Comment by asaage** [comment]
hatte daher noch diesen [Vorschlag ](https://github.com/contao/calendar-bundle/issues/5#issuecomment-255441700) eingebracht...

**Comment by Physiklehrer** [comment]
> hatte daher noch diesen Vorschlag eingebracht...

Danke, dass du dies in Erinnerung rufst. Für diese Realisierung spricht auf jeden Fall, dass die Einstellung genau da vorgenommen wird, wo sie relevant ist (nämlich im Modul EventLISTE).

In den anderen Fällen ("offen" / "geschlossen" als Eigenschaft des Events selbst oder des Kalenders) ist aus Usability-Sicht fragwürdig, dass hier eine Einstellung vorgenommen werden kann, die sich nur auf einen Spezialfall bezieht (nämlich Modul EventLISTE im Modus upcoming) und in allen anderen Konstellationen keine Auswirkungen hat (Kalender, Eventliste im Modus Tag, Jahr, ...).

**Comment by leofeyer** [comment]
Ich habe das bisherige Verhalten in Contao 3 in c82ac67e9ce4c9b9caa98d0161b555957ee5a356 wiederhergestellt.

**Comment by leofeyer** [comment]
Die wählbare Option im Modul "Eventliste" habe ich in contao/calendar-bundle@fd89938aa834f40ae560039893ccde27ff72ee6e hinzugefügt.

<img width="752" alt="" src="https://cloud.githubusercontent.com/assets/1192057/19641558/88d20532-99e1-11e6-8298-ee82140c1e16.png">
